lib/urn.rb

```diff
@@ -1,26 +1,88 @@
+def URN(urn)
+  URN.new(urn)
+end
+
 class URN
+  InvalidURNError = Class.new(StandardError)
+
   PATTERN = %{(?i:urn:(?!urn:)[a-z0-9][a-z0-9\-]{1,31}:} +
-            %{(?:[a-z0-9()+,-.:=@;$_!*']|%[0-9a-f]{2})+)}.freeze
+            %{(?:[a-z0-9()+,-.:=@;$_!*']|%(?:2[1-9a-f]|[3-9a-f][0-9a-f]))+)}.freeze
   REGEX = /\A#{PATTERN}\z/
 
   attr_reader :urn
   private :urn
 
+  def self.extract(str)
+    if block_given?
+      str.scan(/#{PATTERN}/) { yield $& }
+      nil
+    else
+      result = []
+      str.scan(/#{PATTERN}/) { result.push $& }
+      result
+    end
+  end
+
   def initialize(urn)
+    fail InvalidURNError, "bad URN(is not URN?): #{urn}" if urn !~ REGEX
+
     @urn = urn
   end
 
-  def valid?
-    !(urn =~ REGEX).nil?
+  def nid
+    _scheme, nid, _nss = parse
+
+    nid
+  end
+
+  def nss
+    _scheme, _nid, nss = parse
+
+    nss
   end
 
   def normalize
-    return unless valid?
+    _scheme, nid, nss = parse
 
-    _scheme, nid, nss = urn.split(':', 3)
     normalized_nid = nid.downcase
     normalized_nss = nss.gsub(/%([0-9a-f]{2})/i) { |hex| hex.downcase }
 
-    "urn:#{normalized_nid}:#{normalized_nss}"
+    URN.new("urn:#{normalized_nid}:#{normalized_nss}")
+  end
+
+  def to_s
+    urn
+  end
+
+  def ===(other)
+    if other.respond_to?(:normalize)
+      urn_string = other.normalize.to_s
+    else
+      begin
+        urn_string = URN.new(other).normalize.to_s
+      rescue URN::InvalidURNError
+        return false
+      end
+    end
+
+    normalize.to_s == urn_string
+  end
+
+  def ==(other)
+    return false unless other.is_a?(URN)
+
+    normalize.to_s == other.normalize.to_s
+  end
+
+  def eql?(other)
```

Comment:
The work around `==`, `===` and `eql?` is based on [Addressable::URI](https://github.com/sporkmonger/addressable/blob/master/lib/addressable/uri.rb#L2154) implementation.